keps/sig-node/4210-max-image-gc-age/README.md

```diff
@@ -0,0 +1,370 @@
+## Summary
+
+Add an option ImageMaximumGCAge, which allows an admin to specify a time after which unused images will be garbage collected
+by the Kubelet, regardless of disk usage.
+
+## Motivation
+
+Currently, all image garbage collection the Kubelet is triggered by disk usage going over a threshold (ImageGCLowThresholdPercent).
+However, there are cases that additional conditions could be considered useful. One such condition is maximum age of an image.
+If an image is unused for a long time (the exact amount of time will be decided, but on the order of weeks is what comes to mind),
+then it is not likely to be used again.
+
+One such condition that can be imagined are clusters with automatic upgrades. If a cluster goes through an upgrade process,
+it will likely have images cached from the old release (old kube-apiserver/etcd/etc). While these images would eventually get removed
+through the disk usage condition, they will needlessly occupy disk space before that.
+
+### Goals
+
+- Introduce an option to the Kubelet ImageMaximumGCAge
+
+### Non-Goals
+
+- Introduce other conditions for image garbage collection to trigger.
+    - A WG was put together in SIG-Node to collect other GC use cases, and no other conditions were identified.
+
+## Proposal
+
+Kubelet has three different configuration fields for image garbage collection:
+- ImageMinimumGCAge: the youngest an image can be to be qualified for garbage collection
+- ImageGCLowThresholdPercent: The lowest disk usage will be before garbage collection begins
+- ImageGCHighThresholdPercent: The highest disk usage will be before garbage collection runs each GC period.
+
+Between each of these options, there is a common thread: image garbage collection only triggers when disk usage has reached a certain threshold.
+In other words, there are no alternative conditions the Kubelet will begin triggering collection. Luckily, this condition is the most important:
+the primary goal of garbage collection is to ensure images don't clutter the disk too much and cause it to fill up needlessly.
+However, having the Kubelet be purely reactive means that images will clutter the disk and cause it to fill up. While there aren't reported cases
+where this causes issues, it is inefficient with disk space and can cause the Kubelet to scramble to save disk space when the threshold is met.
+
+An additional approach is to define a way for an admin to request images are cleaned up after they're unused for a certain period of time.
+This would reduce the frequency of the disk usage hitting the level, and provide an admin more flexibility in how garbage collection is defined.
+
+The proposal of this KEP is to add an option to the KubeletConfiguration object that looks like:
+```
+	// imageMaximumGCAge is the maximum age an image can be unused before it is garbage collected.
+	ImageMaximumGCAge metav1.Duration
+```
+
+To begin, this option will be set to the maximum duration (math.MaxInt64), which effectively disables.
```

Comment:
Wouldn't it make more sense that if this isn't explicitly set, then just don't run this policy?

That seems to be more aligned with the API way of adding new fields?

Reply:
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version

Reply:
I don't believe there is an unset for it, as metav1.Duration just wraps and marshals a time.Duration, which is just an int64--no pointers involved. Other fields have +optional on, but they default to 0. We could default to 0 in this case, interpreting it as "off" but that would prevent users from setting ImageMaximumGCAge to zero: meaning GC immediately

Reply:
I've decided to set it to 0 by default and add `+optional`, along with adding a note that says 0 will disable the field

Reply:
SGTM.  API review might point out that all new fields should be pointers but I am not sure if that is followed for configurations.  THis would be the case for new API fields.